src/compute/src/render/reduce.rs

```diff
@@ -1123,23 +1145,26 @@ where
     /// stages can skip validation.
     fn build_bucketed_negated_output<S, R>(
         &self,
-        input: &Collection<S, ((Row, u64), Row), Diff>,
+        input: &Collection<S, (Row, Row), Diff>,
         aggrs: Vec<AggregateFunc>,
-    ) -> Collection<S, ((Row, u64), R), Diff>
+    ) -> Collection<S, (Row, R), Diff>
     where
         S: Scope<Timestamp = G::Timestamp>,
-        R: MaybeValidatingRow<Row, (Row, u64)>,
+        R: MaybeValidatingRow<Row, Row>,
     {
         let error_logger = self.error_logger();
         // NOTE(vmarcos): The input operator name below is used in the tuning advice built-in
         // view mz_internal.mz_expected_group_size_advice.
         let arranged_input =
-            input.mz_arrange::<KeyValSpine<_, Row, _, _>>("Arranged MinsMaxesHierarchical input");
+            input.mz_arrange::<RowRowSpine<_, _>>("Arranged MinsMaxesHierarchical input");
 
         arranged_input
-            .mz_reduce_abelian::<_, KeyValSpine<_, _, _, _>>(
+            .mz_reduce_abelian::<_, RowValSpine<_, _, _>>(
```

Comment:
Note that the output arrangement for a stage is only evolved to a `RowValSpine`. This is because we may want to encode either `Result<Row, Row>` or `Row` in the value to detect invalid accumulations early in the first stage of a hierarchical reduction. This is less error prone, since we report the invalid accumulation as soon as we see it instead of delaying the reporting to another operator, and allows us to not render a reduce pair.

Reply:
Commit e119c10db8c81b7f28e2d238cec831543dfef998 in a sequenced PR shows how to improve the above to only keep to `RowValSpine` for the first level of the arrangement stack, and use `RowRowSpine` for all remaining levels.